### PR TITLE
Adding support and examples for AOSS

### LINKED ISSUE #204: [FEATURE] Add support for OpenSearch Serverless
### Is your feature request related to a problem?

- OpenSearch serverless service identifier changes from `es` to `aoss`.
- Sigv4 requires a hash
- Info API doesn't exist
- What else?

### What solution would you like?

Document support for OpenSearch Serverless in the compatibility matrix, add integration tests.

### Do you have any additional context?

Coming from https://github.com/opensearch-project/opensearch-clients/issues/44

### PR DESCRIPTION
**Pull request by VachaShah**
Signed-off-by: Vacha Shah <vachshah@amazon.com>

### Description
Adding support and examples for AOSS.

### Issues Resolved
#204 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by dblock** [comment]
I fully expect this not to be needed soon for AOSS, this is work in progress server-side.

**Comment by dblock** [comment]
The root API will be (re)added to OpenSearch Serverless.

I propose we remove the product check altogether, similar to https://github.com/opensearch-project/opensearch-ruby/issues/130. We try to do a version check to see if the client is compatible with the server. I believe this was inherited from Elastic that was trying to maintain some sense of version compatibility without semver, but since we follow semver it's easy to know now whether your client will work against a server or not. My concern is that a version request every single time a new instance of a client is created is not insignificant overhead (in a web app example this could end up making a root request every time a web app serves a client search request!), and is an unexpected side effect when users try to use the client.

**Comment by VachaShah** [comment]
> I fully expect this not to be needed soon for AOSS, this is work in progress server-side.

Sounds good! I removed the content-length change from this PR.

**Comment by VachaShah** [comment]
> The root API will be (re)added to OpenSearch Serverless.
> 
> I propose we remove the product check altogether, similar to [opensearch-project/opensearch-ruby#130](https://github.com/opensearch-project/opensearch-ruby/issues/130). We try to do a version check to see if the client is compatible with the server. I believe this was inherited from Elastic that was trying to maintain some sense of version compatibility without semver, but since we follow semver it's easy to know now whether your client will work against a server or not. My concern is that a version request every single time a new instance of a client is created is not insignificant overhead (in a web app example this could end up making a root request every time a web app serves a client search request!), and is an unexpected side effect when users try to use the client.

Sure! I removed the product info check altogether.

**Comment by VachaShah** [comment]
@dblock This is ready for review.

File: USER_GUIDE.md [hunk 1]
```diff
@@ -1,6 +1,9 @@
 - [User Guide](#user-guide)
-    - [Example](#example)
-    - [How to use IAMs as authentication method](#how-to-use-iams-as-authentication-method)
+	- [Example](#example)
+	- [How to use IAMs as authentication method](#how-to-use-iams-as-authentication-method)
+			- [AWS SDK V1](#aws-sdk-v1)
+			- [AWS SDK V2](#aws-sdk-v2)
+			- [AWS SDK V2 with OpenSearch Serverless](#aws-sdk-v2-with-opensearch-serverless)
```
Comment (dblock) [inline USER_GUIDE.md:6]: Tabs vs. spaces? 🤔

File: USER_GUIDE.md [hunk 2]
```diff
@@ -283,3 +286,91 @@ func main() {
 }
 
 ```
+
+#### AWS SDK V2 with OpenSearch Serverless
```
Comment (dblock) [inline USER_GUIDE.md:290]: We should merge the managed and serverless portions, these are the same now no?

**Review by dblock — CHANGES_REQUESTED** [review]
I think we should make a separate PR with an entry that removes the version check, so we can make it very visible. Should we major increment version, too?

For docs I'd say managed and serverless work the same and the only thing you want to change is the service signing name. Let's say that in the user guide?

File: USER_GUIDE.md (hunk 1 above)
Comment (VachaShah) [reply USER_GUIDE.md]: My bad, fixed

File: USER_GUIDE.md (hunk 2 above)
Comment (VachaShah) [reply USER_GUIDE.md]: Combined them.

**Comment by VachaShah** [comment]
> I think we should make a separate PR with an entry that removes the version check, so we can make it very visible. Should we major increment version, too?
> 
> For docs I'd say managed and serverless work the same and the only thing you want to change is the service signing name. Let's say that in the user guide?

Created a separate PR #219 for removing the version check and incrementing the version. 
Updated the user guide.

File: USER_GUIDE.md [hunk 3]
```diff
@@ -1,6 +1,8 @@
 - [User Guide](#user-guide)
-    - [Example](#example)
-    - [How to use IAMs as authentication method](#how-to-use-iams-as-authentication-method)
```
Comment (dblock) [inline USER_GUIDE.md:3]: I know we already had this here, but shouldn't it be "IAM"?

More generally I think we might want to align this on other docs and say "Amazon OpenSearch Service" as a title.

File: USER_GUIDE.md [hunk 4]
```diff
@@ -221,32 +223,39 @@ func main() {
 
 #### AWS SDK V2
 
+AWS SDK V2 for Go can be used for various service options. To create a client, please see the example below:
```
Comment (dblock) [inline USER_GUIDE.md:226]: I don't think we need to be polite in docs, would remove "please".

I also prefer succinct versions, so "Use the AWS SDK v2 for Go to authenticate with Amazon OpenSearch service."

File: USER_GUIDE.md [hunk 5]
```diff
@@ -221,32 +223,39 @@ func main() {
...
+	"github.com/aws/aws-sdk-go-v2/aws"
 	"github.com/aws/aws-sdk-go-v2/config"
-	"github.com/opensearch-project/opensearch-go/v2/opensearchapi"
+	opensearch "github.com/opensearch-project/opensearch-go/v2"
+	opensearchapi "github.com/opensearch-project/opensearch-go/v2/opensearchapi"
 	requestsigner "github.com/opensearch-project/opensearch-go/v2/signer/awsv2"
 )
 
-const endpoint = "" // e.g. https://opensearch-domain.region.com
+const endpoint = "" // e.g. https://opensearch-domain.region.com or serverless collection endpoint
```
Comment (dblock) [inline USER_GUIDE.md:242]: Naming-wise we call it Amazon OpenSearch Serverless

File: USER_GUIDE.md [hunk 6]
```diff
@@ -260,26 +269,51 @@ func main() {
 		log.Fatal("client creation err", err)
 	}
 
-	ping := opensearchapi.PingRequest{}
+	indexName = "go-test-index"
 
-	resp, err := ping.Do(ctx, client)
+    // Define index mapping.
```
Comment (dblock) [inline USER_GUIDE.md:274]: align

**Review by dblock — APPROVED** [review]
Looks good! I have some potential nits for docs, feel free to ignore them and merge.

**Comment by VachaShah** [comment]
@dblock I addressed the comments for documentation, it needs your approval once more.

**Review by dblock — APPROVED** [review]

File: USER_GUIDE.md [hunk 7]
```diff
@@ -260,26 +269,51 @@ func main() {
...
 
-	resp, err := ping.Do(ctx, client)
+	// Define index mapping.
+	mapping := strings.NewReader(`{
+	 "settings": {
+	   "index": {
+	        "number_of_shards": 4
+	        }
+	      }
+	 }`)
```
Comment (VijayanB) [inline USER_GUIDE.md:281]: Where is this mapping used?

File: signer/awsv2/sdkv2signer.go [hunk 8]
```diff
@@ -68,6 +68,7 @@ func (s *awsSdkV2Signer) SignRequest(r *http.Request) error {
 	}
 
 	hash, err := hexEncodedSha256OfRequest(r)
+	r.Header.Set("X-Amz-Content-Sha256", hash)
```
Comment (VijayanB) [inline signer/awsv2/sdkv2signer.go:71]: Shall we move this to constant?

File: USER_GUIDE.md (hunk 7 above)
Comment (VachaShah) [reply USER_GUIDE.md:281]: My bad, I added its usage

File: signer/awsv2/sdkv2signer.go (hunk 8 above)
Comment (VachaShah) [reply signer/awsv2/sdkv2signer.go:71]: Since its just used here, I kept it as a string

**Comment by VachaShah** [comment]
@VijayanB @dblock Review again please?

**Review by dblock — APPROVED** [review]

**Pull request merged**